sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveReferencesInAggregate.scala

```diff
@@ -0,0 +1,202 @@
+ *    `ResolveLateralColumnAliasReference` will further resolve [[LateralColumnAliasReference]] and
+ *    rewrite the plan. This is to support SQL pattern like
+ *    `SELECT col1 + 1 AS x, x + 1 AS y, y + 1 AS z FROM t`.
+ * 5. Resolves the column to outer references with the outer plan if we are resolving subquery
+ *    expressions.
+ */
+object ResolveReferencesInAggregate extends SQLConfHelper with ColumnResolutionHelper {
+  def apply(a: Aggregate): Aggregate = {
+    val planForResolve = a.child match {
+      // SPARK-25942: Resolves aggregate expressions with `AppendColumns`'s children, instead of
+      // `AppendColumns`, because `AppendColumns`'s serializer might produce conflict attribute
+      // names leading to ambiguous references exception.
+      case appendColumns: AppendColumns => appendColumns
+      case _ => a
+    }
+
+    val resolvedAggExprsNoOuter = a.aggregateExpressions
+      .map(resolveExpressionByPlanChildren(_, planForResolve, allowOuter = false)
+        .asInstanceOf[NamedExpression])
+
+    val resolvedGroupingExprs = a.groupingExpressions
+      .map(resolveExpressionByPlanChildren(_, planForResolve, allowOuter = false))
+      // SPARK-31670: Resolve Struct field in groupByExpressions and aggregateExpressions
+      // with CUBE/ROLLUP will be wrapped with alias like Alias(GetStructField, name) with
+      // different ExprId. This cause aggregateExpressions can't be replaced by expanded
+      // groupByExpressions in `ResolveGroupingAnalytics.constructAggregateExprs()`, we trim
+      // unnecessary alias of GetStructField here.
+      .map(trimTopLevelGetStructFieldAlias)
+
+    // Only makes sense to do the rewrite once all the aggregate expressions have been resolved.
+    // Otherwise, we might incorrectly pull an actual aggregate expression over to the grouping
+    // expression list (because we don't know they would be aggregate expressions until resolved).
+    if (resolvedAggExprsNoOuter.forall(_.resolved)) {
+      val finalGroupExprs = resolveGroupByAll(
+        resolvedAggExprsNoOuter,
+        resolveGroupByAlias(resolvedAggExprsNoOuter, resolvedGroupingExprs)
+      ).map(resolveOuterRef)
+      a.copy(finalGroupExprs, resolvedAggExprsNoOuter, a.child)
+    } else {
+      // If the SELECT list is not full resolved at this point, we need to apply lateral column
+      // alias and outer reference resolution, which are not supported in GROUP BY. We can't
+      // resolve group by alias and group by all here.
+      // Aggregate supports Lateral column alias, which has higher priority than outer reference.
+      val resolvedAggExprsWithLCA = resolveLateralColumnAlias(resolvedAggExprsNoOuter)
+      val resolvedAggExprsWithOuter = resolvedAggExprsWithLCA.map(resolveOuterRef)
+        .map(_.asInstanceOf[NamedExpression])
+      a.copy(resolvedGroupingExprs.map(resolveOuterRef), resolvedAggExprsWithOuter, a.child)
```

Comment:
It can be problematic, try with this query:
```
      sql(s"select * from values(1) as t(something) " +
        s"where exists (select salary * 1.5 as something from $testTable group by something)")
```
where testTable has a salary field. (I just wrote in LCA suite.)

Reply:
Good catch! It's important to remember that rules will be executed repeatedly and we need to make sure the rule won't conflict with its previous runs.